### PR TITLE
Sms retry limit

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Will not allow a person to retry a sms activation code more than three times
## Impacts

Changes the workflow of sms activation
## Delete branch after merge

NO

### CODE REVIEW

File: html/captive-portal/lib/captiveportal.pm [hunk 1]
```diff
@@ -111,8 +112,16 @@ sub loadCustomStatic {
...
+=head2 user_cache
+
+Returns the user specific cache
+
+=cut
+
 sub user_cache {
-    return pf::CHI->new( namespace => 'httpd.portal');
+    my ($c) = @_;
+    #Return a CHI::Driver that prefixes each
```
Comment (dwlfrth) [inline html/captive-portal/lib/captiveportal.pm]: each... ? :P

File: html/captive-portal/lib/captiveportal/Base/Controller.pm [hunk 2]
```diff
@@ -44,6 +44,24 @@ sub showError {
...
+
+=cut
+
+sub reached_retry_limit {
+    my ( $self, $c, $retry_key, $max ) = @_;
+    my $cache = $c->user_cache;
+    my $retries = $cache->get($retry_key) || 0;
+    $retries++;
+    if($retries > $max ) {
+        return 1;
```
Comment (dwlfrth) [inline html/captive-portal/lib/captiveportal/Base/Controller.pm:60]: In these case, I'd prefer using $TRUE and $FALSE...

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm [hunk 3]
```diff
@@ -119,22 +119,30 @@ sub showSmsConfirmation : Private {
...
     # no form was submitted, assume first time
     my $pin = $c->request->param("pin");
     if ($pin) {
-        $c->log->info("Mobile phone number validation attempt");
-        if ( pf::activation::validate_code($pin) ) {
-            return ( $TRUE, 0 );
-        } else {
-            return ( $FALSE, $GUEST::ERROR_INVALID_PIN );
+        $logger->info("Mobile phone number validation attempt");
+        if ($self->reached_retry_limit($c, 'sms_retries', 3)) {
```
Comment (dwlfrth) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm]: Should be a configuration parameter (of the source)
- Do we enforce limiting
- Limit

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm [hunk 4]
```diff
@@ -119,22 +119,30 @@ sub showSmsConfirmation : Private {
...
+            my $mac = $c->portalSession->clientMac;
+            $logger->info("Max tries reached invalidating code for $mac");
```
Comment (dwlfrth) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm:133]: I'd add the limit (in this case, 3) in the comment

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm (hunk 3 above)
Comment (extrafu) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm]: s/Should be/Must be/

We must never hardcode things like that in the code, unless if a default and safe value if no configuration is found.

**Comment by jrouzierinverse** [comment]
Please Review

File: html/captive-portal/lib/captiveportal/Base/Controller.pm [hunk 5]
```diff
@@ -44,6 +44,26 @@ sub showError {
...
+=head2 reached_retry_limit
+
+Test if the retry limit has been reached for a session key
+If the max is undef or 0 then it will return true
+
+=cut
+
+sub reached_retry_limit {
+    my ( $self, $c, $retry_key, $max ) = @_;
+    return 1 unless $max;
```
Comment (cgx) [inline html/captive-portal/lib/captiveportal/Base/Controller.pm]: If you set _sms_pin_retry_limit_ to 0 (as in `conf/profiles.conf.example`), the limit will always be reached and the user will never be able to validate her phone number. Is this what we want? If the limit is set to 0, I would expect the limit to be disabled.

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm [hunk 6]
```diff
@@ -119,22 +119,32 @@ sub showSmsConfirmation : Private {
...
 
     # no form was submitted, assume first time
     my $pin = $c->request->param("pin");
     if ($pin) {
-        $c->log->info("Mobile phone number validation attempt");
-        if ( pf::activation::validate_code($pin) ) {
-            return ( $TRUE, 0 );
-        } else {
-            return ( $FALSE, $GUEST::ERROR_INVALID_PIN );
+        $logger->info("Mobile phone number validation attempt");
```
Comment (cgx) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Sms.pm]: I would categorize this log entry into _debug_.

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 7]
```diff
@@ -241,6 +241,19 @@ has_field 'nbregpages' =>
...
+
+=cut
+
+has_field 'sms_pin_retry_limit' =>
+  (
+    type => 'PosInteger',
+    label => 'SMS Pin Retry Limit',
+    default => 0,
+    tags => { after_element => \&help,
+             help => 'How many times a user can retry an sms pin having to request anothor pin. A value of 0 mean disables retries check.' },
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: I would rephrase this: 

_Maximum number of times a user can retry a SMS PIN before having to request another PIN. A value of 0 disables the limit._

File: lib/pf/activation.pm [hunk 8]
```diff
@@ -303,6 +307,39 @@ sub invalidate_codes {
     return;
 }
 
+=item invalidate_codes_for_mac
```
Comment (cgx) [inline lib/pf/activation.pm]: Should be `=head2`, no?

File: lib/pf/web/guest.pm [hunk 9]
```diff
@@ -490,6 +491,7 @@ Readonly::Hash our %ERRORS => (
     $ERROR_SPONSOR_NOT_ALLOWED  => 'Email %s is not allowed to sponsor guest access',
     $ERROR_PREREG_NOT_ALLOWED  => 'Guest pre-registration is not allowed by policy',
     $ERROR_INVALID_PIN => 'PIN is Invalid!',
+    $ERROR_MAX_RETRIES => 'Maximum amount of retries attempted',
```
Comment (cgx) [inline lib/pf/web/guest.pm:494]: Please update `conf/locale/en/LC_MESSAGES/packetfence.po`. I will eventually extend `extract_i18n_strings.pl` to cover the captive portal and not only the Web admin.

File: lib/pf/activation.pm (hunk 8 above)
Comment (rouzier) [inline lib/pf/activation.pm]: Yes I will update the entire pod

File: lib/pf/web/guest.pm (hunk 9 above)
Comment (rouzier) [inline lib/pf/web/guest.pm:494]: Done
